### PR TITLE
build: remove cyclic dependencies

### PR DESCRIPTION
**Pull request by antjkennedy**
* restructure code to get rid of cyclic dependencies
* add cyclic dependency check to tslint
* add aztec example

Signed-off-by: Ant Kennedy <ant@antkennedy.com>

### CODE REVIEW

**Comment by antjkennedy** [comment]
Tests do not pass yet, but restructured the code to remove the circular dependencies so hopefully in a better position.

Still needs a lot of work and tidying up before ready for production

File: src/test/core/aztec/detector/Detector.spec.ts [hunk 1]
```diff
@@ -208,11 +208,11 @@ describe('DetectorTest', () => {
     }
 
     function getOrientationPoints(code: AztecCode): Point[] {
-        let center: int = code.getMatrix().getWidth() / 2;
-        let offset: int = code.isCompact() ? 5 : 7;
+        let center: number = code.getMatrix().getWidth() / 2;
```
Comment (odahcam) [inline src/test/core/aztec/detector/Detector.spec.ts]: Why changing `int` for `number`?

**Comment by odahcam** [comment]
What you're doing here is great, but we gotta keep the code aligned with the Java so we can keep porting their codebase with ease. If you let me ask something, I would rather prefer passing tests for new code than changes in the original API, do you think it's possible to reduce the API changes so we can accept this faster?

Also, I'm not saying that we will never change the original APIs, I find it a bad piece of code myself, but it's a little problematic to do it now.

If you still wanna break the cyclic dependencies I would ask you to at least keep the code in the same file, other class, but same file you could even create pure functions for some methods.

**Comment by odahcam** [comment]
Also, shouldn't we finish porting Aztec before going into this?

**Comment by antjkennedy** [comment]
Will have a look at how I can reduce the code changes, it might be that this is a no go due to current code structure.

**Comment by odahcam** [comment]
Okay, I'll wait for updates then. Again, I like the changes, but I'm obligated to keep this codebase somehow the closest to the Java version; anyway you can always have multiple class implementations in a TS file, since in TS files are known as modules and it's ok to put different  things in it.

**Pull request merged**